Allow parameters for todo.sh default action

chrysle opened the pull request:
**Before submitting a pull request,** please make sure the following is done:

- [x] Fork [the repository](https://github.com/todotxt/todo.txt-cli) and create your branch from `master`.
- [x] If you've added code that should be tested, add tests!
- [x] Ensure the test suite passes.
- [x] Format your code with [ShellCheck](https://www.shellcheck.net/).
- [x] Include a human-readable description of what the pull request is trying to accomplish.
- [ ] Steps for the reviewer(s) on how they can manually QA the changes.
- [ ] Have a `fixes #XX` reference to the issue that this pull request fixes.

This finishes leftover work from #159 
That PR can be closed then.

I added a test, a docstring and also a note for `TODOTXT_DEFAULT_ACTION` to the config file.

butonic committed:
allow parameters / options in default action

chrysle committed:
Added test and docstring

chrysle committed:
Added `TODOTXT_DEFAULT_ACTION` to the config

inkarkat committed:
Actually test (and fix) that the default action parameter is received by the action

The foo2 action needs to print its parameters (minus the first which is the action name, instead of hard-coding "foo" if no parameters are passed (which also is broken because the $# is already expanded in the here-doc)), and the test needs to include "foo" (instead of "bar") to match the (reused) expected output.

inkarkat commented:
Unfortunately, you're written the test to match the implementation, not to actually verify the specified behavior; and that actually was broken (good that we insisted on having tests). I've corrected both in my commit. It still doesn't handle default parameters that have spaces or other special characters; I'll add that later.

chrysle commented:
Thanks for your help! What special parameters for example do you mean that should be tested?

chrysle commented:
@inkarkat I would like to relieve you of implementing the tests if ou give me some hints.

inkarkat committed:
Also handle special characters (that need escaping) in the default action

Spaces, quotes, $, and other special shell characters need to be escaped inside the TODOTXT_DEFAULT_ACTION, and consequently eval() used to parse that back into individual arguments.

inkarkat committed:
FIX: Default action with arguments does not consider built-in commands

Only $TODO_ACTIONS_DIR/$action, not even the $TODO_ACTIONS_DIR/$action/$action variant.
A totally different approach is needed: Instead of parsing off the first argument and treating that as the custom action, do a recursive invocation with the default action parsed as a command-line. A new isDefaultAction flag is needed so that we only do this when the default action actually applies (otherwise, strange invocations like todo.sh "add +project here" would also be possible). The condition for non-empty $TODOTXT_DEFAULT_ACTION avoids an endless loop.

chrysle commented on tests/t0002-actions.sh:
```diff
@@ -37,4 +44,41 @@ test_expect_success 'custom action (env variable)' '
 
+test_expect_success 'custom action (default action)' '
+    mkdir .todo.actions.d
+    cp foo2 .todo.actions.d/
```
Maybe use the logical `&&` operator here too?

inkarkat replied:
Yes, makes sense.

chrysle commented on tests/t0002-actions.sh:
```diff
@@ -16,6 +16,13 @@ echo "TODO: foo"
 
+cat > foo2 << 'EOF'
+shift
+IFS=-
```
Why did you put dashes between the output?

inkarkat replied:
To be sure that the individual arguments are passed separately, and no wrong splitting on contained spaces occurs. I've added a comment to document this.

chrysle commented on todo.sh:
```diff
@@ -727,7 +727,13 @@ if [ -n "$OVR_TODOTXT_FINAL_FILTER" ] ; then
+    ACTION=$1
+else
+    ACTION=$TODOTXT_DEFAULT_ACTION
+    isDefaultAction=t
```
Maybe use `true` here to improve readability?

inkarkat replied:
The `test-lib.sh` already uses the short `t` for truth, and it's a common idiom.

chrysle reviewed (commented):
I have left some comments.

inkarkat committed:
Robustness: Use mkdir -p and copy only on success

inkarkat commented:
Thanks for the review @chrysle; the original implementation was woefully incomplete and wrong, but only writing the tests showed what looked quite alright to a casual review. From my point of view this is now done, and we just have to wait for the merge of the huge backlog of open issues...

chrysle commented:
Did you notice the test for macOS 10.15 succeeded? I wonder why it was always cancelled before.

inkarkat committed:
Merge branch 'master' into allow-params-default-action

inkarkat committed:
Documentation: Update changelog

Closes #159

inkarkat committed:
todo.cfg: Cosmetics: Don't overload "parameters" (use "arguments"), drop trailing spaces
